### ISSUE TITLE
Class property reflection issues

### ISSUE THREAD

**Issue by m1waterman88** [bitpay/php-bitpay-client-v2#192]
We're seeing the following errors after upgrading from 6.0.2111 to 7.1.0, on line 673 in `JsonMapper::createInstance`. It's failing on what starts as a `Client::getInvoice` call using the merchant facade. Our code hasn't changed.

```
Class "\BitPaySDK\Model\Invoice\$overpaidAmount" not found
```

```
Class "\BitPaySDK\Model\Invoice\$underpaidAmount" not found
```

I'm thinking there's a reflection issue in there since these are properties---or seemingly method parameters given the missing underscore---we don't use on our side, and it appears the SDK code has getters and setters for it without actually using them outside of tests.

bitpay/sdk: 7.1.0
PHP: 8.1

Can you let me know if there's something we need to change on our end that you can surmise from this? Of course, let me know if you need any more information.

**Comment by bobbrodie** [comment]
Hi there @m1waterman88 ! We will take a look at this tonight, thank you so much for reporting.

**Comment by bobbrodie** [comment]
@m1waterman88 we're checking this out. JsonMapper checks types both in the objects and the annotations. In 7.1.0, the annotation doesn't have a type specified, so we added it in one of the recent pull requests that cleans up documentation, so in `master` we have tagged each as `int`:

```php
    /**
     * Sets overpaidAmount
     *
     * This parameter will be returned on the invoice object
     * if the invoice was overpaid ("exceptionStatus": "paidOver").
     * It equals to the absolute difference between amountPaid
     * and paymentTotals for the corresponding transactionCurrency used.
     *
     * @param int $overpaidAmount the overpaid amount
     */
    public function setOverpaidAmount($overpaidAmount)
    {
        $this->_overpaidAmount = $overpaidAmount;
    }
```

And:
```php
    /**
     * Sets underpaidAmount
     *
     * This parameter will be returned on the invoice object
     * if the invoice was underpaid ("exceptionStatus": "paidPartial").
     * It equals to the absolute difference between amountPaid
     * and paymentTotals for the corresponding transactionCurrency used.
     *
     * @param int $underpaidAmount the underpaid amount
     */
    public function setUnderpaidAmount($underpaidAmount)
    {
        $this->_underpaidAmount = $underpaidAmount;
    }
```

We are checking the rest of the annotations and running a few tests before we do a release of 7.2.0. If you could, would you be able to share a simple replication method? That'll ensure we're covering your scenario. Thanks!

**Comment by doliveira-v** [comment]
Hey @bobbrodie thanks for checking this so quickly, the information we pass to the sdk is really just an invoice ID (not sure if they're safe to share, so I cropped the image) to bitpay's `getInvoice` method and no other parameters.

<img width="869" alt="Screenshot 2022-12-29 at 10 52 59 AM" src="https://user-images.githubusercontent.com/121567700/209981477-e8f12433-2990-4e3a-baf4-fb38f0655458.png">

**Comment by bobbrodie** [comment]
Wonderful, thank you so much! We’re testing a change and will keep everyone posted 🙂

**Comment by bobbrodie** [comment]
Hi @m1waterman88 & @doliveira-v! we're running through some manual testing and wanted to check if you could share an example of a JSON response (just the non-sensitive data) you've received from the API that throws the error. Thanks!

**Comment by doliveira-v** [comment]
@bobbrodie please let me know if that's enough information

<img width="359" alt="Screenshot 2022-12-30 at 9 15 41 AM" src="https://user-images.githubusercontent.com/121567700/210080402-6d0c7f49-fa34-4e37-a030-4265c62b504b.png">
<img width="873" alt="Screenshot 2022-12-30 at 9 20 27 AM" src="https://user-images.githubusercontent.com/121567700/210080406-48e9943e-8cf2-47ad-8be9-453d3c1c05d7.png">

**Comment by bobbrodie** [comment]
Thank you so much @doliveira-v, this is where it gets interesting. The response doesn't have `overpaidAmount` or `underpaidAmount` in it, which is typical. We've tested creating an invoice and getting an invoice with PHP 8.1 and 8.2 with SDK 7.1.0 and so far have been unsuccessful in replicating (even in the new version we're building that has integration tests).

That said, we've adjusted the annotations in master, so if you could try the following locally or in a test environment (some non-disruptive place) we'd be grateful:

1. Update your `composer.json` to use `dev-master` for the SDK version
2. Just to be safe, remove your `vendor/` directory and `composer.lock` file and run `composer install` **OR** run `composer update bitpay/sdk`
3. Re-run your test

Thanks again for all the feedback - since we're not in your application it is extremely helpful!

**Comment by doliveira-v** [comment]
@bobbrodie alright, I see your point, we're setting up test cases in our dev environment currently, in the meanwhile can you please provide your next estimated release date for 7.2?

**Comment by doliveira-v** [comment]
Hi @bobbrodie just checking, are we on track for 7.2 this week or would you say most likely next week?

**Comment by bobbrodie** [comment]
Hi @doliveira-v we are working on the release this week, but would like to confirm if testing against `dev-master` resolved your issue? We haven't been able to replicate, so we want to ensure that the changes have addressed it for you. Thanks!

**Comment by doliveira-v** [comment]
@bobbrodie thanks for replying, test assertions work fine in local for `dev-master` , but we can't upgrade production to a dependency flagged as `dev-master` maybe @m1waterman88 can be able to reproduce the error somehow in a unit test in his local/dev but ideally we would just like to update to v7.2 and hopefully that'd fix the issue in prod.

**Comment by bobbrodie** [comment]
Hey everyone, we've released 7.2.0. We'll close this for now, but please let us know if we need to reopen. Thanks!
